### DISCUSSION TITLE
Constant BPM mode

### DISCUSSION THREAD

**Discussion by rasputozen** [ppy/osu#15428] [Ideas]
This mode would automatically adjust the halftime/daycore/doubletime speed sliders to keep a chosen BPM constant across all songs.

**Comment by frenzibyte** [comment]
Can you elaborate further on this mod? is it a mod that allows the user to choose a certain BPM and the game will automatically adjust the speed in relation to the beatmap's original BPM? (and therefore either slow it down or speed it up?)

**Reply by rasputozen** [reply]
Yeah that's exactly what I mean.

**Reply by mk56-spn** [reply]
how would  this work on songs with  multiple bpms ? does it speed up and slow down each segment to keep it consistent?    if so this sounds really hard to implement

**Reply by rasputozen** [reply]
I think for that case a practical solution would be to adjust the whole song according to whatever its longest-duration BPM is. That way you still get a meaningful BPM shift while maintaining the integrity of the song.

**Reply by RichardKanshen** [reply]
having all segments have same bpm as an option would not be so bad as well

**Reply by takoz53** [reply]
[osu trainer](https://github.com/FunOrange/osu-trainer) is implementing this already and could be taken as a reference to do this. This is done by making use of the "average bpm", also visible in osu! client on songs.

I haven't dug into his code a lot, but logically speaking, this approach could be viable:
Assuming we want the next Song to be 240 BPM:
If a song ranges with eg. 200, 210, 220 and has 210 as average:
210 will be pushed to 240. Calculating 240/210, this is a 1.14x multiplier.

Using this multiplier, we can just do 200 * 1.14 and 220 * 1.14 as well. Respectively, the slower parts will be 228 BPM and the faster parts will be 250 BPM.

I brought this up, as I find the current implementation, the multiplier-slider, really unintuitive and would rather set the BPM by typing in a number. If I want to train for 240 BPM songs, I don't want to drag the slider back and forth to find out which multiplier now really makes the song be 240 BPM. Doing this once is all good and fine, but when you want to do this throughout a session, it can get really, really annoying! :)

What I would rather expect is the user typing in their BPM value and the Slider adjusting itself after the user has inputted their value.
